Broadcom iProc (ARM Cortex-A9) PMU driver
Sourced from Broadcom LDK 3.4.10

diff --git a/drivers/bcmdrivers/pmu/.gitignore b/drivers/bcmdrivers/pmu/.gitignore
new file mode 100644
index 0000000..af7303c
--- /dev/null
+++ b/drivers/bcmdrivers/pmu/.gitignore
@@ -0,0 +1,8 @@
+/.built-in.o.cmd
+/.iproc_pmu.o.cmd
+/.iproc-pmu.o.cmd
+/built-in.o
+/iproc_pmu.o
+/iproc-pmu.o
+/modules.builtin
+/modules.order
diff --git a/drivers/bcmdrivers/pmu/Kconfig b/drivers/bcmdrivers/pmu/Kconfig
new file mode 100644
index 0000000..8b09c0c
--- /dev/null
+++ b/drivers/bcmdrivers/pmu/Kconfig
@@ -0,0 +1,9 @@
+
+config IPROC_PMU
+	tristate "Iproc PMU support"
+	depends on ARCH_IPROC
+	default y
+	help
+	  Iproc PMU support
+	  This config provides kernel-side support for iProc PMU description.
+	  Its generic driver is perf_event.c
diff --git a/drivers/bcmdrivers/pmu/Makefile b/drivers/bcmdrivers/pmu/Makefile
new file mode 100644
index 0000000..b491218
--- /dev/null
+++ b/drivers/bcmdrivers/pmu/Makefile
@@ -0,0 +1,8 @@
+# iproc TDM support
+
+EXTRA_CFLAGS	+= -I$(BCMDRIVERS_DIR)/gmac/src/include
+KBUILD_CFLAGS += -DBCMDRIVER -Dlinux
+
+iproc_pmu-objs := iproc-pmu.o
+
+obj-$(CONFIG_IPROC_PMU) += iproc_pmu.o
diff --git a/drivers/bcmdrivers/pmu/iproc-pmu.c b/drivers/bcmdrivers/pmu/iproc-pmu.c
new file mode 100644
index 0000000..83e1aec
--- /dev/null
+++ b/drivers/bcmdrivers/pmu/iproc-pmu.c
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2013, Broadcom Corporation. All Rights Reserved.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * PMU device description to Iproc
+ */
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <typedefs.h>
+#include <osl.h>
+#include <bcmutils.h>
+#include <mach/reg_utils.h>
+#include <asm/pmu.h>
+#include <mach/io_map.h>
+
+static struct resource iproc_pmu_resource = {
+    .start  = BCM_INT_ID_IHOST_PMU,
+    .end    = BCM_INT_ID_IHOST_PMU+1,
+    .flags  = IORESOURCE_IRQ,
+};
+
+
+static struct platform_device iproc_pmu_device = {
+	.name		=	"arm-pmu",
+	.id		    =	ARM_PMU_DEVICE_CPU,
+	.dev =  {
+		.init_name = "arm-pmu",
+	},
+        .num_resources  = 1,
+        .resource = &iproc_pmu_resource,
+};
+
+
+static int __init iproc_pmu_init(void)
+{
+    int ret;
+    printk(KERN_INFO "Registering iproc_pmu_device\n");
+    ret = platform_device_register(&iproc_pmu_device);
+    return ret;
+}
+module_init(iproc_pmu_init);
+
+
+/* Module information */
+MODULE_DESCRIPTION("IPROC PMU Driver");
+MODULE_LICENSE("GPL");
